libraries/AP_Scripting/lua_bindings.cpp

```diff
@@ -58,10 +60,172 @@ static const luaL_Reg servo_functions[] =
     {NULL, NULL}
 };
 
+static int AP_Logger_Write(lua_State *L) {
+    AP_Logger * AP_logger = AP_Logger::get_singleton();
+    if (AP_logger == nullptr) {
+        return luaL_argerror(L, 1, "logger not supported on this firmware");
+    }
+
+    // check we have at least 4 arguments passed in
+    const int args = lua_gettop(L);
+    if (args < 4) {
+        return luaL_argerror(L, args, "too few arguments");
+    }
+
+    const char * name = luaL_checkstring(L, 1);
+    const char * labels = luaL_checkstring(L, 2);
+    const char * fmt = luaL_checkstring(L, 3);
+
+    // cheack the name, labels and format are not too long
+    if (strlen(name) >= LS_NAME_SIZE) {
+        return luaL_error(L, "Name must be less than 5 chars long");
+    }
+    if (strlen(labels) >= (LS_LABELS_SIZE - 7)) { // need 7 chars to add 'TimeUS,'
+        return luaL_error(L, "labels must be less than 58 chars long");
+    }
+    int length = strlen(fmt);
+    if (length >= (LS_FORMAT_SIZE - 1)) { // need 1 chars to add timestamp
+        return luaL_error(L, "format must be less than 16 chars long");
+    }
+
+    // check the number of arguments matches the length of the foramt string
+    if (args - 3 != length) {
+        return luaL_argerror(L, args, "format does not match No. of arguments");
+    }
+
+    // prepend timestamp to format and labels
+    char label_cat[LS_LABELS_SIZE];
+    strcpy(label_cat,"TimeUS,");
+    strcat(label_cat,labels);
+    char fmt_cat[LS_FORMAT_SIZE];
+    strcpy(fmt_cat,"Q");
+    strcat(fmt_cat,fmt);
+    length++;
+
+    // ask for a mesage type
+    struct AP_Logger::log_write_fmt *f = AP_logger->msg_fmt_for_name(name, label_cat, nullptr, nullptr, fmt_cat);
+    if (f == nullptr) {
+        // unable to map name to a messagetype; could be out of
+        // msgtypes, could be out of slots, ...
+        return luaL_argerror(L, args, "could not map message type");
+    }
+
+    // work out how long the block will be
+    int16_t msg_len = AP_logger->Write_calc_msg_len(fmt_cat);
+    if (msg_len == -1) {
+        return luaL_argerror(L, args, "unknown format");
+    }
+
+    const int32_t buffer_size = msg_len * sizeof(uint8_t);
+    uint8_t * buffer = (uint8_t *) hal.util->heap_realloc(lua_scripts::_heap, nullptr, buffer_size);
+    if (buffer == nullptr) {
+        return luaL_error(L, "Could not allocate log buffer");
+    }
+
+    // zero out buffer
+    for (int32_t ii=0; ii<buffer_size; ii++) {
+        buffer[ii] = 0;
```

Comment:
`memset`. Definetly `memset` not a loop.

Reply:
fixed